crates/sui-indexer/src/models/tx_indices.rs

```diff
@@ -101,37 +114,68 @@ impl TxIndex {
                 cp_sequence_number,
                 tx_sequence_number,
                 recipient: s.to_vec(),
+                transaction_kind,
             })
             .collect();
-        let tx_input_objects = self
-            .input_objects
-            .iter()
-            .map(|o| StoredTxInputObject {
-                cp_sequence_number,
-                tx_sequence_number,
-                object_id: bcs::to_bytes(&o).unwrap(),
-            })
-            .collect();
-        let tx_changed_objects = self
-            .changed_objects
-            .iter()
-            .map(|o| StoredTxChangedObject {
-                cp_sequence_number,
-                tx_sequence_number,
-                object_id: bcs::to_bytes(&o).unwrap(),
-            })
-            .collect();
-        let tx_calls = self
-            .move_calls
-            .iter()
-            .map(|(p, m, f)| StoredTxCalls {
-                cp_sequence_number,
-                tx_sequence_number,
-                package: p.to_vec(),
-                module: m.to_string(),
-                func: f.to_string(),
-            })
-            .collect();
+
+        let mut address_rel_map: HashMap<&SuiAddress, i16> = HashMap::new();
+        for sender in &self.senders {
+            address_rel_map.entry(sender).or_insert(0);
+        }
+
+        for recipient in &self.recipients {
+            address_rel_map
+                .entry(recipient)
+                .and_modify(|e| {
+                    if *e == 0 {
+                        *e = 1
+                    }
+                })
+                .or_insert(2);
+        }
+
+        let mut tx_input_objects = Vec::new();
+
+        for tx_input_object in self.input_objects {
+            for (address, rel) in &address_rel_map {
```

Comment:
I am worried about the storage size here, if we now have a txn with 1 sender, 1000 recipients and 1000 input objects, we will have roughly 1 million rows for this txn in `tx_input_objects` table, this is very in-efficient, also we do not prune on these indices tables, which can be very problematic.

Reply:
meanwhile if I just do JOIN with cursors, the perf seems fine to me
```
EXPLAIN ANALYZE SELECT io.tx_sequence_number
FROM tx_input_objects io
JOIN tx_senders s ON io.tx_sequence_number = s.tx_sequence_number
WHERE io.object_id = decode('66bfc923a016e87080d51712837fa16c706b76de9ca3282f03a2268f2cd7cee8', 'hex') AND io.tx_sequence_number > 16155
AND s.sender = decode('7d20dcdb2bca4f508ea9613994683eb4e76e9c4ed371169677c1be02aaf0b58e', 'hex') AND s.tx_sequence_number > 16155 LIMIT 100;

 Limit  (cost=1.40..662.80 rows=100 width=8) (actual time=0.024..0.384 rows=100 loops=1)
   ->  Nested Loop  (cost=1.40..715298.75 rows=108150 width=8) (actual time=0.023..0.377 rows=100 loops=1)
         ->  Index Only Scan using tx_input_objects_pkey on tx_input_objects io  (cost=0.70..50868.45 rows=118453 width=8) (actual time=0.012..0.025 rows=100 loops=1)
               Index Cond: ((object_id = '\x66bfc923a016e87080d51712837fa16c706b76de9ca3282f03a2268f2cd7cee8'::bytea) AND (tx_sequence_number > 16155))
               Heap Fetches: 0
         ->  Index Only Scan using tx_senders_pkey on tx_senders s  (cost=0.70..5.61 rows=1 width=8) (actual time=0.003..0.003 rows=1 loops=100)
               Index Cond: ((sender = '\x7d20dcdb2bca4f508ea9613994683eb4e76e9c4ed371169677c1be02aaf0b58e'::bytea) AND (tx_sequence_number = io.tx_sequence_number) AND (tx_sequence_number > 16155))
               Heap Fetches: 0
 Planning Time: 0.222 ms
 Execution Time: 0.406 ms
```

Reply:
I prob missed some analyses / experiments you have done, let's huddle on this.

Reply:
IIRC @wlmyng has run migrations on one of the dbs. What is the resulting size of the input/changed obj tables?

Reply:
```
defaultdb=> select pg_relation_size('tx_input_objects');
 pg_relation_size
------------------
     502932111360
(1 row)

defaultdb=> select pg_relation_size('tx_input_objects_rel');
 pg_relation_size
------------------
     784261488640
(1 row)

defaultdb=> select pg_relation_size('tx_changed_objects');
 pg_relation_size
------------------
     533476925440
(1 row)

defaultdb=> select pg_relation_size('tx_changed_objects_rel');
 pg_relation_size
------------------
    1958039273472
(1 row)
```

looks like it's tx_changed_objects that suffers the most

Reply:
yeah it's substantial amount of storage, our current tnt DB is about 7TB and mnt is about 17TB last time I checked.

Reply:
Actually doesn't this blowup apply to other tables too? not just input objects and changed objects.

Reply:
defaultdb=> select pg_relation_size('tx_input_objects');
 pg_relation_size
------------------
     502932111360
(1 row)

defaultdb=> select pg_relation_size('tx_input_objects_rel');
 pg_relation_size
------------------
     784261488640
(1 row)

defaultdb=> select pg_relation_size('tx_changed_objects');
 pg_relation_size
------------------
     533476925440
(1 row)

defaultdb=> select pg_relation_size('tx_changed_objects_rel');
 pg_relation_size
------------------
    1958039273472
(1 row)

defaultdb=> select pg_relation_size('tx_calls');
 pg_relation_size
------------------
     158479106048
(1 row)

defaultdb=> select pg_relation_size('tx_calls_rel');
 pg_relation_size
------------------
     226317180928